### DISCUSSION TITLE
Architecture Discussion

### DISCUSSION THREAD

**Discussion by DandyLyons** [DandyLyons/React-Example#5] [General]
In this discussion, I would like us to work toward making decisions regarding code organization and frameworks used. The goal is to keep the code as consistent, simple, scalable, and testable as possible.

 In many cases, there are many framework that have already been created to solve these problems. For example, I'm interested in using the [Effect](https://effect.website/) library because it makes it easier to write safe code that handles errors. But it also comes with a little overhead of learning a new library. This is a tradeoff. In other words, I don't want to add a framework unless I'm confident that it will reduce more complexity then it will add. 
 
In this discussion, please share any frameworks, tools, best practices etc. that we should consider using in this project.

**Comment by therealjustinbiele** [comment]
### Routing
A huge part of client-side development is page routing. What happens if you web application becomes complex enough to where you need multiple pages to handle certain content on the site? Take airbnb for example: they have a home page, signup page, account details page, a property/reservation page. 

[React Router DOM](https://reactrouter.com/en/main/start/tutorial) is sort of the de-facto library for handling client-side routing in React apps.

This library allows developers to define a `Router` component; this component acts as sort of the root `App` component we're familiar with. The Router component accepts a `routes` prop, which is just an array of Route definitions: objects with a "path" and "element" key. 

Route Interface
```
interface Route {
  path: string;
  element: JSX.Element
}
``` 
So you could define a route like this:

```
import AboutPage from './AboutPage'

const routeDefinitions = [
  { path: "/about", element: <AboutPage /> }
];

<Router routes={routeDefinitions}>
```

**Reply by therealjustinbiele** [reply]
This is also helpful for collaborating with others; ie, you work on the Profile page while someone else works on the Product List page

**Reply by DandyLyons** [reply]
React Router DOM sounds like a prime candidate. The architecture I'm interested in for this program is battle tested solutions with a long history and lots of documentation. If it's considered a "de facto" library then that's definitely something I want to consider using.

**Reply by therealjustinbiele** [reply]
I'll get a POC committed to a branch and open a PR 👍

**Comment by therealjustinbiele** [comment]
+1 on effect libraries, though it's always fun to write your own ;)

**Comment by DandyLyons** [comment]
@therealjustinbiele What are you thoughts on Redux. I've recently learned "The Composable Architecture", a Swift library which is heavily inspired by Redux. Personally, I like the pattern. I think it makes projects much easier to reason about and prevent bugs. But it does require more boilerplate which can be overkill for smaller projects. Do you feel the same way about Redux? I have no direct experience with Redux, just TCA. 

Also, for this project, I would like to keep it small and concise. Beginner friendly.
